Fix DB lock errors, clean up build tags

lstoll opened the pull request:
Add a test to make reproducing the concurrency behaviour easier.

Remove shared cache mode. This interacting with how go manages database connections seems to be the cause of our locking errors: https://github.com/mattn/go-sqlite3/issues/632#issue-360437494. Instead, just use a busy timeout and let sqlite figure it out.

Remove sqlite build tags 

* libsqlite3 no longer needed, no more errors on OS X with spatialite using the default config
* unlock_notify didn't fix our errors, now doing concurrency at the app level

lstoll committed:
Add failing concurrency test, fix with RWMutex

Add a test to make reproducing the concurrency behaviour easier.

Turns out sqlite allows one write, _or_ multiple readers. Deal with this by using a per-table
RWMutex to handle this behvaiour at the app level.

lstoll committed:
Remove sqlite build tags

libsqlite3 no longer needed, no more errors on OS X with spatialite using the default config

unlock_notify didn't fix our errors, now doing concurrency at the app level

lstoll committed:
bang on concurrency a little more

alindeman commented on sql.go:
```diff
@@ -313,9 +313,11 @@ var migrations = []migration{
-	// go-sqlite supports concurrent reads, but not writes. Queries that write
-	// should use this mutex to synchronize that access
-	writeMu sync.Mutex
+	// sqlite supports either one writer, or multiple readers. wrap tables in
```
TIL

lstoll replied:
I'm still not 100% sure where the concurrency boundary is, I can find a lot of takes on various threads and issues but no real succinct proof.

I did manage to confirm that an insert while a select is running does fail with a table lock tho.

lstoll replied:
Aaah this is a little insightful: https://github.com/mattn/go-sqlite3/issues/632#issue-360437494

Seems like it's not the database concurrency in https://sqlite.org/lockingv3.html we're hitting, it's actually queries being interleaved on the same connection.

After reading through that it seems like the shared cache is not really making things better, and a busy timeout would be a more appropriate solution. Gonna give that a shot.

alindeman commented on sql_test.go:
```diff
@@ -42,12 +45,94 @@ func TestMigrations(t *testing.T) {
+
+			for i := 0; i < 10; i++ {
+				if err := s.AddOTLocation(ctx, otMsg); err != nil {
+					errs = append(errs, err)
```
I'm pretty sure it's not safe to `append` in multiple goroutines for the same slice (unless using a mutex, of course). I usually use a `chan error` for this kind of thing.

lstoll committed:
Don't use shared cache mode, use a busy timeout

Turns out shared cache mode limits db connections, which ultimate triggers a lot of the locking issues:
https://github.com/mattn/go-sqlite3/issues/632#issue-360437494

Stop using shared cache mode, in favour of a busy timeout. This lets us remove the mutexes, and
appears to solve the issue.

lstoll committed:
test uses spatialite driver too

lstoll committed:
simplify test again, use channels for errs

lstoll commented:
Reworked based on new learnings

lstoll commented:
#32 replaces this
